### PR TITLE
plan: reuse expressionRewriter.

### PR DESCRIPTION
**Pull request by winoros**
Mainly to optimize prepare && execute statement's cost.
Benchmark result is generated by https://github.com/pingcap/tidb-lightning/tree/cenqichao/kv-encoder-bench
At master `bb2440b63690c75e1be23da6995cf00924e9658c`:
```
[stmt] kv num = 24576000 / kv size = 2724864000
[stmt] cost avg = 71.807 sec
```

This pr:
```
[stmt] kv num = 24576000 / kv size = 2724864000
[stmt] cost avg = 62.593 sec
```

The non-prepare's cost of two branch is all around 66s.

PTAL @shenli @coocood @hanfei1991

### CODE REVIEW

File: plan/planbuilder.go [hunk 1]
```diff
@@ -164,6 +164,9 @@ type planBuilder struct {
 	optFlag       uint64
 
 	curClause clauseCode
+
+	rewriterStack  []*expressionRewriter
+	rwStackPointer int
```
Comment (hanfei1991) [inline plan/planbuilder.go]: add comment for the both fields

File: plan/planbuilder.go [hunk 2]
```diff
@@ -164,6 +164,9 @@ type planBuilder struct {
 	optFlag       uint64
 
 	curClause clauseCode
+
+	rewriterStack  []*expressionRewriter
```
Comment (hanfei1991) [inline plan/planbuilder.go]: rewriterPool is better ?

File: plan/expression_rewriter.go [hunk 3]
```diff
@@ -67,13 +67,26 @@ func (b *planBuilder) rewrite(expr ast.ExprNode, p LogicalPlan, aggMapper map[*a
...
 	expression.Expression, LogicalPlan, error) {
-	er := &expressionRewriter{
-		p:          p,
-		aggrMap:    aggMapper,
-		b:          b,
-		asScalar:   asScalar,
-		ctx:        b.ctx,
-		preprocess: preprocess,
+	b.rwStackPointer++
+	var er *expressionRewriter
```
Comment (coocood) [inline plan/expression_rewriter.go]: `er` is too similar to the short name of `error`.

File: plan/planbuilder.go (hunk 1 above)
Comment (coocood) [inline plan/planbuilder.go]: `rw` or `rewriter`?
Should be consistent.

File: plan/expression_rewriter.go [hunk 4]
```diff
@@ -67,13 +67,26 @@ func (b *planBuilder) rewrite(expr ast.ExprNode, p LogicalPlan, aggMapper map[*a
...
+			p:          p,
+			aggrMap:    aggMapper,
+			b:          b,
+			asScalar:   asScalar,
+			ctx:        b.ctx,
+			preprocess: preprocess,
+		}
+		b.rewriterStack = append(b.rewriterStack, er)
+	} else {
+		// If er.err is not nil, the planner will fail and won't continue. So don't need to set the er.err's value.
```
Comment (shenli) [inline plan/expression_rewriter.go]: Can not understand it.

File: plan/expression_rewriter.go [hunk 5]
```diff
@@ -67,13 +67,26 @@ func (b *planBuilder) rewrite(expr ast.ExprNode, p LogicalPlan, aggMapper map[*a
...
+	if len(b.rewriterStack) < b.rwStackPointer {
+		er = &expressionRewriter{
+			p:          p,
+			aggrMap:    aggMapper,
+			b:          b,
+			asScalar:   asScalar,
+			ctx:        b.ctx,
+			preprocess: preprocess,
+		}
+		b.rewriterStack = append(b.rewriterStack, er)
```
Comment (shenli) [inline plan/expression_rewriter.go]: Will this cause too many growslice in some scenario?

Comment (winoros) [reply plan/expression_rewriter.go]: In most cases, there won't be many rewriters.

**Comment by coocood** [comment]
LGTM

**Pull request merged**